src/client/gpm_common.c

```diff
@@ -300,55 +299,41 @@ static int gpm_epoll_wait(struct gpm_ctx *gpmctx, uint32_t event_flags)
-            /* We got an event which was not EPOLLIN; assume this is an error,
-             * and exit with EBADF: epoll_wait said timerfd had an event,
-             * but that event is not an EPOLIN event. */
-            ret = EBADF;
-        } else {
-            ret = read(gpmctx->timerfd, &timer_read, sizeof(uint64_t));
-            if (ret == -1 && errno != EAGAIN && errno != EWOULDBLOCK) {
-                /* In the case when reading from the timer failed, don't hide the
-                 * timer error behind ETIMEDOUT such that it isn't retried */
-                ret = errno;
-            } else {
-                /* If ret == 0, then we definitely timed out. Else, if ret == -1
-                 * and errno == EAGAIN or errno == EWOULDBLOCK, we're in a weird
-                 * edge case where epoll thinks the timer can be read, but it
-                 * is blocking more; treat it like a TIMEOUT and retry, as
-                 * nothing around us would handle EAGAIN from timer and retry
-                 * it. */
-                ret = ETIMEDOUT;
-            }
-        }
-        gpm_epoll_close(gpmctx);
-    } else {
-        /* If ret == 2, then we ignore the timerfd; that way if the next
-         * operation cannot be performed immediately, we timeout and retry.
-         * Always check the returned event of the socket fd. */
-        int fd_index = 0;
-        if (epoll_ret == 2 && events[fd_index].data.fd != gpmctx->fd) {
-            fd_index = 1;
+        goto done;
+    } else if (epoll_ret == 2 || events[0].data.fd == gpmctx->timerfd ||
+               events[1].data.fd == gpmctx->timerfd) {
+        /* Timer fired - service it. */
+        ret = read(gpmctx->timerfd, &timer_read, sizeof(uint64_t));
+        if (ret == -1 && errno != EAGAIN && errno != EWOULDBLOCK) {
+            /* In the case when reading from the timer failed, don't hide the
+             * timer error behind ETIMEDOUT such that it isn't retried */
+            ret = errno;
+            goto done;
         }
 
-        if ((events[fd_index].events & event_flags) == 0) {
-            /* We cannot call EPOLLIN/EPOLLOUT at this time; assume that this
-             * is a fatal error; return with EBADFD to distinguish from
-             * EBADF in timer_fd case. */
-            ret = EBADFD;
-            gpm_epoll_close(gpmctx);
-        } else {
-            /* We definintely got a EPOLLIN/EPOLLOUT event; return success. */
-            ret = 0;
-        }
+        /* If ret == 0, then we definitely timed out. Else, if ret == -1 and
+         * errno == EAGAIN or errno == EWOULDBLOCK, we're in a weird edge case
+         * where epoll thinks the timer can be read, but it is blocking more;
+         * treat it like a TIMEOUT and retry, as nothing around us would
+         * handle EAGAIN from timer and retry it. */
+        ret = ETIMEDOUT;
+        goto done;
     }
 
+    if ((events[0].events & event_flags) == 0) {
```

Comment:
I don't understand this part. Or it is overly tricksy... I think `events[0]` is wrong here...? What tells us this is the `gpmctx->fd` and not `gpmctx->timerfd`? We care about the former, not the latter... I guess in the event `ret == 2`, because one of the two FDs is the `timerfd`, we end up going to `done`, rather than getting here... So this is tricky an warrants a comment at minimum? :-)

See, this isn't just a refactor, this is also a behavior change. In the old code, we checked this _prior_ to checking/reading `timerfd`, which means that, with your always exit (new line 323 above), we wouldn't trigger this and would return `ETIMEDOUT` instead...

Reply:
The check you pointed out had the wrong conditional handles all cases where the timer fires.  At that point, we know that the timer cannot have fired; there are only two fds, so events[0] is for gpmctx->fd.

And yes, its a behavior change.  You're commenting on the bugfix commit :)